### PR TITLE
Istio operator blog

### PR DESCRIPTION
**Pull request by frankbu**

### CODE REVIEW

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+Kubernetes [operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) provide
+a pattern for encoding human operations knowledge in software and are a popular way to simplify
+the administration of software infrastructure components.
+Because Istio is complex to administer, it's a natural candidate for an operator.
+
+Up until now, [Helm](https://github.com/helm/helm) has been the primary tool to install and upgrade Istio.
+In addition to needing to install another tool, this approach is not so desireable because it lacks
+schematic validation, allowing invalid configurations to be applied, which can then only be detected
+after installation. With operators, on the other hand, the API is schematized in a
+CRD (Custom Resource Definition) which is validated.
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: Helm users will take exception to this since it's in principle possible to validate the schema (though not conveniently). There are other reasons for getting rid of helm which are explained in the design doc but for the blog I suggest we leave this section out and not say anything negative about Helm.

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+{{< text bash >}}
+$ kubectl apply -f https://<repo URL>/<new version>/operator-profile-default.yaml
+{{< /text >}}
+
+Refer to [standalone operator](/docs/ops/setup/standalone-operator/) for more details.
+
+## Using the {{< istioctl >}} CLI
+
+In the meantime, until the controller is thouroughly tested and better integrated with `istioctl`,
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: Actually I think controller is not likely to become the preferred method of install and admin for security reasons. I think it's better to steer users to istioctl first in the blog.

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    $ helm template ... --set global.mtls.enabled=true
+    {{< /text >}}
+
+    You can use this `istioctl` command:
+
+    {{< text bash >}}
+    $ istioctl manifest generate ... --set values.global.mtls.enabled=true
+    {{< /text >}}
+
+1. Automated migration from helm `values.yaml` using the `istioctl manifest migrate` command (alpha in Istio 1.4).
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: Link to the command in istio.io?

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+1. The full helm installation API can also be accessed as a pass-through from `istioctl` and controller.
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: This point is essentially saying the same thing as the first point. You could combine them for example like this:
```
The full helm installation API can also be accessed as a pass-through from `istioctl` and controller.  Pass Helm configuration options using `istioctl --set` by prepending the string “values.“  to the option name. For example, instead of this helm command:
---
example
---
 See [Customize Istio settings using the Helm API](/docs/setup/install/operator/#customize-istio-settings-using-the-helm-api)
￼   for details. 
```

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 3 above)
Comment (frankbu) [reply content/en/blog/2019/introducing-istio-operator/index.md]: good idea

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 1 above)
Comment (frankbu) [reply content/en/blog/2019/introducing-istio-operator/index.md]: We can tone it down to sound less negative about Helm, but I think it's very important that this blog explains why we're replacing it. It's a significant disruption for users, so we need to be clear about why the new approach is better.

I guess another benefit of operator is the easy continuous incremental update provided by the controller. Can you point me to the where in the design doc reasons are explained? I'll take another pass at justifying the change more diplomatically.

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+in favor of a new operator-based [installation using {{< istioctl >}}](/docs/setup/install/operator/).
+Upgrading from Istio 1.4 onward (that is, versions not initially installed with Helm)
+will also be done using a new [{{< istioctl >}} upgrade feature](/docs/setup/upgrade/istioctl-upgrade/).
+
+## Istio Control Plane API
+
+An operator implementation requires a CRD (Custom Resource Definition) to define the API.
+The Istio operator CRD is defined by the [`IstioControlPlane` API](/docs/reference/config/istio.operator.v1alpha12.pb/).
+This API supports all of Istio's current [configuration profiles](/docs/setup/additional-setup/config-profiles/)
+using a single field to select the profile. For example, the following CRD configures Istio using the `demo` profile:
```
Comment (sbezverk) [inline content/en/blog/2019/introducing-istio-operator/index.md]: I think here you are referring to CR custom resource and not CRD which is just defines a structure of a custom resource. The instantiation of CRD is CR, which is confirmed by the output below.

Comment (frankbu) [reply content/en/blog/2019/introducing-istio-operator/index.md]: Thanks, good point although, unlike CRD, I don't think CR is a recognized acronym, so I'll just reword it to `the following IstioControlPlane resource configures Istio ...`

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 1 above)
Comment (ostromart) [reply content/en/blog/2019/introducing-istio-operator/index.md]: How about dropping the above paragraph and adding something along the lines of the following, after the next paragraph:
"The new istioctl CLI and operator controller both validate configuration schemas and perform a range of checks for installation change or upgrade. These checks are tightly integrated with the tools to prevent common errors and simplify troubleshooting."
The fact is we did have a lot of problems with Helm, but I'd rather not dwell on it. Maybe we can say something more vague like "increased robustness"...

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+a pattern for encoding human operations knowledge in software and are a popular way to simplify
+the administration of software infrastructure components.
+Because Istio is complex to administer, it's a natural candidate for an operator.
+
+Up until now, [Helm](https://github.com/helm/helm) has been the primary tool to install and upgrade Istio.
+This approach worked well but had some disadvantages:
+
+1. Users need to install another tool.
+1. Supporting Istio's many small specializations cause Helm templates to become unwieldy.
+1. Helm configurations are difficult to validate using Istio's proto-based programming model.
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: "...are difficult to validate for Istio's proto-based APIs." ?
Also, maybe:
1. Upgrades sometimes require Istio specific hooks which have been difficult to implement and maintain with Helm.

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+With operators, the API is schematized in a
+[Custom Resource Definition](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions)
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: schematized in a [CustomResourceDefinition]() which is generated from a [proto](/docs/reference/config/istio.operator.v1alpha12.pb/)

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+which is more easily managed and validated at runtime.
+
+Starting with Istio 1.4, the [Helm installation](/docs/setup/install/helm/) approach has been deprecated
+in favor of a new operator-based [installation using {{< istioctl >}}](/docs/setup/install/operator/).
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: ...in favor of istioctl-based, which shares the CRD with operator.

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+Upgrading from Istio 1.4 onward (that is, versions not initially installed with Helm)
+will also be done using a new [{{< istioctl >}} upgrade feature](/docs/setup/upgrade/istioctl-upgrade/).
+
+The newly released Istio operator simplifies the common administrative tasks of installation,
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: The newly released istioctl commands and operator simplifies... ?

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  name: example-istiocontrolplane
+spec:
+  profile: demo
+  telemetry:
+    enabled: false
+{{< /text >}}
+
+## Using the {{< istioctl >}} CLI
+
+The preferred way to use the Istio operator API is through a new set of `istioctl` CLI commands.
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: recommended way?

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+{{< text bash >}}
+$ istioctl manifest generate ... --set values.global.mtls.enabled=true
+{{< /text >}}
+
+You can also set Helm configuration values in an `IstioControlPlane` configuration resource.
+See [Customize Istio settings using the Helm API](/docs/setup/install/operator/#customize-istio-settings-using-the-helm-api)
+￼for details. 
+
+Another feature to help with the transition is the
```
Comment (ostromart) [inline content/en/blog/2019/introducing-istio-operator/index.md]: ...transition is the alpha [manifest migrate]

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
+---
+title: Introducing the Istio Operator
+description: Introduction to Istio's new operator-based installation and control plane management feature.
+publishdate: 2019-11-05
```
Comment (frankbu) [inline content/en/blog/2019/introducing-istio-operator/index.md]:
```suggestion
publishdate: 2019-11-12
```

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+Up until now, [Helm](https://github.com/helm/helm) has been the primary tool to install and upgrade Istio.
+This approach worked well but had some disadvantages:
+
+1. Users need to install another tool.
+1. Supporting Istio's many small specializations cause Helm templates to become unwieldy.
+1. Helm configurations are difficult to validate using Istio's proto-based APIs.
+1. Upgrades sometimes require Istio specific hooks that have been difficult to implement and maintain with Helm.
+
+Starting with Istio 1.4, the [Helm installation](/docs/setup/install/helm/) approach has been deprecated
+in favor of a new [installation using {{< istioctl >}}](/docs/setup/install/operator/) approach.
```
Comment (frankbu) [inline content/en/blog/2019/introducing-istio-operator/index.md]:
```suggestion
in favor of a new [installation using {{< istioctl >}}](/docs/setup/install/istioctl/) approach.
```

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+the `istioctl` commands can also be passed individual settings using a `--set` flag:
+
+{{< text bash >}}
+$ istioctl manifest apply --set telemetry.enabled=false
+{{< /text >}}
+
+There are also a number of other `istioctl` commands that, for example,
+help you list, display, and compare configuration profiles and manifests.
+
+Refer to the Istio [install instructions](/docs/setup/install/operator) for more details.
```
Comment (frankbu) [inline content/en/blog/2019/introducing-istio-operator/index.md]:
```suggestion
Refer to the Istio [install instructions](/docs/setup/install/istioctl) for more details.
```

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+{{< /text >}}
+
+You can use this `istioctl` command:
+
+{{< text bash >}}
+$ istioctl manifest generate ... --set values.global.mtls.enabled=true
+{{< /text >}}
+
+You can also set Helm configuration values in an `IstioControlPlane` configuration resource.
+See [Customize Istio settings using Helm](/docs/setup/install/operator/#customize-istio-settings-using-the-helm-api)
```
Comment (frankbu) [inline content/en/blog/2019/introducing-istio-operator/index.md]:
```suggestion
See [Customize Istio settings using Helm](/docs/setup/install/istioctl/#customize-istio-settings-using-the-helm-api)
```

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+
+Up until now, [Helm](https://github.com/helm/helm) has been the primary tool to install and upgrade Istio.
+This approach worked well but had some disadvantages:
+
+1. Users need to install another tool.
+1. Supporting Istio's many small specializations cause Helm templates to become unwieldy.
+1. Helm configurations are difficult to validate using Istio's proto-based APIs.
+1. Upgrades sometimes require Istio specific hooks that have been difficult to implement and maintain with Helm.
+
+Starting with Istio 1.4, the [Helm installation](/docs/setup/install/helm/) approach has been deprecated
```
Comment (howardjohn) [inline content/en/blog/2019/introducing-istio-operator/index.md]: What does deprecated mean? How long will we continue to support Helm?

Migrating from Helm to the operator is going to be a huge change for existing users. We should assure them exactly how long they have to do this migration.

@ostromart mentioned we are not recommending existing users migrate from helm to the operator. Is this the case? Should we express that here?

Comment (frankbu) [reply content/en/blog/2019/introducing-istio-operator/index.md]: Since [Helm is a Beta feature](https://istio.io/about/feature-stages/#core), the standard 3 months [deprecation policy](https://istio.io/about/feature-stages/) applies. I think that goes without saying, but we could mention it in the release notes.

Comment (howardjohn) [reply content/en/blog/2019/introducing-istio-operator/index.md]: That would imply we would delete Helm in 1.5 which I think is not the plan, so we should probably call it out here?

Comment (ssuchter) [reply content/en/blog/2019/introducing-istio-operator/index.md]: I kind of agree with John. I think we should use the word "recommended" instead of "deprecated" (e.g. Starting with Istio 1.4, a new installation using istioctl is recommended, instead of the Helm installation approach.) Deprecated has a **very specific** documentation meaning that seems likely to be interpreted the way John suggests. We should only use it if we intend to actually follow its meaning.

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+in favor of a new [installation using {{< istioctl >}}](/docs/setup/install/istioctl/) approach.
+Upgrading from Istio 1.4 onward (that is, versions not initially installed with Helm)
+will also be done using a new [{{< istioctl >}} upgrade feature](/docs/setup/upgrade/istioctl-upgrade/).
+
+The new `istioctl` commands use a
+[Custom Resource Definition (CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions)
```
Comment (ssuchter) [inline content/en/blog/2019/introducing-istio-operator/index.md]: Should this say Custom Resource Definition or just Custom Resource?

File: content/en/blog/2019/introducing-istio-operator/index.md [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+to configure the installation.
+The CRD is part of a new Istio operator implementation intended to simplify the common administrative tasks of
+installation, upgrade, and complex configuration changes for Istio.
+Validation and checking for installation and upgrade is tightly integrated with the tools to prevent
+common errors and simplify troubleshooting.
+
```
Comment (ssuchter) [inline content/en/blog/2019/introducing-istio-operator/index.md]: I wonder if we could insert a counterpoint section to the helm disadvantages section here.

## Istio Operator benefits

This new CR approach has several benefits:
1. Users just install one tool, Istio Operator, and it handles everything else
1. Users only need to specify the fields they want to customize in their IstioControlPlane resource
1. The Kubernetes API server automatically validates the schema's usage
1. The Istio Operator is a natural place to allow potentially complex Istio upgrade logic and insulate the user from this complexity

**Comment by ssuchter** [comment]
Martin/Frank - since this is already merged, instead of commenting on this PR, should I just make a new one with my suggested changes?

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 16 above)
Comment (ostromart) [reply content/en/blog/2019/introducing-istio-operator/index.md]: Yes, let's not use the word deprecate externally, we'll not be deleting Helm in 3 months. We want to encourage users to switch but migration tools are still alpha. I think we could say "plan to deprecate" or just "recommend istioctl" is probably better for now.

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 17 above)
Comment (ostromart) [reply content/en/blog/2019/introducing-istio-operator/index.md]: Yes, good point. Custom Resource here, but can be left as CRD below.

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 18 above)
Comment (ostromart) [reply content/en/blog/2019/introducing-istio-operator/index.md]: agreed, I think it's better to avoid talking negatively about helm. It has some issues but it's also been a good tool for us.

**Comment by ostromart** [comment]
Yes, please create another PR with edits. Thanks!

**Comment by frankbu** [comment]
@ssuchter That would be great if you submit a PR with your suggested edits. Note that this blog is just one place we need to soften the wording. There are several other places in the docs that mention that helm is being deprecated that also need to be reworded.

One thing I want to be sure of, however, is that we make it clear that the plan is to deprecate helm in the future, so it should be very clear that we don't want people to use the helm instructions. One of the biggest issues with Istio in the past is that there are too many installation instructions, so people don't know which one to use. Nobody wants to read several different install instructions and then try to figure out how to get started. So, the goal is to get down to one instruction in the future and the plan is that istioctl operator will be it. Let's try to make sure not to water down the warning to the point that people don't believe helm will ever be deprecated.

File: content/en/blog/2019/introducing-istio-operator/index.md (hunk 18 above)
Comment (frankbu) [reply content/en/blog/2019/introducing-istio-operator/index.md]: I don't think a new section is warranted, but I do think it would be fine to change the current wording in the overview to list the reasons for switching to operator as benefits instead of helm disadvantages.
